feat(automation): Implement E2E purchase flow with API response handling

alvaroprado199103 opened the pull request:
This commit introduces comprehensive enhancements to the UI automation framework, focusing on the implementation of Phase 3: End-to-End Purchase Flow. The changes include:

1. API Response Handling:
   - Introduced ApiWaitUtils utility class to manage asynchronous API responses
   - Implemented waitForApiResponse method to ensure UI elements are updated after API calls
   - Added proper error handling for API response waiting

2. Code Optimization:
   - Removed unused methods from CartPage and HomePage classes
   - Eliminated redundant code in PracticeSoftwareTestingSteps
   - Standardized WebElement locators using data-test attributes for improved maintainability

3. E2E Purchase Flow Implementation:
   - Enhanced product selection methods with multiple search strategies: * Direct search bar queries * Category-based filtering * Combined filtering with price range and sorting
   - Implemented robust checkout process with proper validation
   - Added invoice verification to confirm successful order association

4. Technical Improvements:
   - Replaced magic numbers with named constants
   - Improved exception handling throughout the codebase
   - Enhanced WebDriverWaitUtils with standardized timeout values
   - Implemented proper cleanup in tearDown methods

These changes ensure reliable execution of the E2E purchase flow, with proper handling of asynchronous operations and validation of the complete user journey from product selection to invoice verification.

alvaroprado199103 committed:
feat(automation): Implement E2E purchase flow with API response handling

This commit introduces comprehensive enhancements to the UI automation framework, focusing on the implementation of Phase 3: End-to-End Purchase Flow. The changes include:

1. API Response Handling:
   - Introduced ApiWaitUtils utility class to manage asynchronous API responses
   - Implemented waitForApiResponse method to ensure UI elements are updated after API calls
   - Added proper error handling for API response waiting

2. Code Optimization:
   - Removed unused methods from CartPage and HomePage classes
   - Eliminated redundant code in PracticeSoftwareTestingSteps
   - Standardized WebElement locators using data-test attributes for improved maintainability

3. E2E Purchase Flow Implementation:
   - Enhanced product selection methods with multiple search strategies:
     * Direct search bar queries
     * Category-based filtering
     * Combined filtering with price range and sorting
   - Implemented robust checkout process with proper validation
   - Added invoice verification to confirm successful order association

4. Technical Improvements:
   - Replaced magic numbers with named constants
   - Improved exception handling throughout the codebase
   - Enhanced WebDriverWaitUtils with standardized timeout values
   - Implemented proper cleanup in tearDown methods

These changes ensure reliable execution of the E2E purchase flow, with proper handling of asynchronous operations and validation of the complete user journey from product selection to invoice verification.

holgiosalos commented on src/test/java/pages/BasePage.java:
```diff
@@ -1,15 +1,57 @@
     protected WebDriver driver;
+    protected WebDriverWait wait;
+    protected WebDriverWaitUtils waitUtils;
+    private static final int DEFAULT_TIMEOUT_SECONDS = 10;
+    private static final Logger LOGGER = Logger.getLogger(BasePage.class.getName());
```
static variables should be declared at the beginning of the class

alvaroprado199103 replied:
Fixed

holgiosalos commented on src/test/java/pages/practicesoftwaretestingpages/AccountPage.java:
```diff
@@ -0,0 +1,27 @@
+    @FindBy(xpath = "//h1[normalize-space()='My account']")
+    private WebElement accountTitle;
+
+    public AccountPage(WebDriver driver, WebDriverWaitUtils waitUtils) {
```
as you are initializing the waitUtils in BasePage, you don't need to declare it as parameter in the constructor of the sub-classes

alvaroprado199103 replied:
Fixed

alvaroprado199103 committed:
This commit introduces comprehensive enhancements to the UI automation framework across multiple phases:

Phase 3: End-to-End Purchase Flow
- Implemented ApiWaitUtils for handling asynchronous API responses
- Enhanced product selection with multiple search strategies (direct search, category filters, combined filters)
- Developed robust checkout process with shipping and payment validation
- Added invoice verification for order confirmation
- Replaced magic numbers with named constants
- Improved exception handling and WebDriverWaitUtils

Phase 4: Allure Reporting Integration
- Implemented comprehensive Allure reporting with detailed step descriptions and severity levels
- Added screenshots for critical test points and detailed attachments for each test step
- Configured Allure for CI/CD with GitHub Actions integration
- Set up retry mechanism for flaky tests and Allure history tracking
- Enhanced test documentation with proper annotations (@Epic, @Feature, @Description)

These changes ensure reliable execution of the E2E purchase flow with proper validation and comprehensive reporting capabilities.

holgiosalos commented on README.md:
Optional change: Review the changes in this file. It seems that the title for section "Current Status and Recent Changes" is duplicated, and phase 2 description was removed

alvaroprado199103 replied:
Fixed

holgiosalos commented on .github/workflows/gradle.yml:
fix the EOF warning
![image](https://github.com/user-attachments/assets/6bc236ce-539e-43d3-bc5a-afcdc61a8ffe)

alvaroprado199103 replied:
Fixed.

holgiosalos commented on src/test/java/pages/practicesoftwaretestingpages/HomePage.java:
```diff
@@ -0,0 +1,135 @@
+    @FindBy(xpath = "//a[@data-test='nav-my-invoices']")
+    private WebElement myInvoicesButton;
+
+    public HomePage(WebDriver driver, WebDriverWaitUtils waitUtils) {
```
as you are initializing the waitUtils in BasePage, you don't need to declare it as parameter in the constructor of the sub-classes

alvaroprado199103 replied:
Fixed.

holgiosalos commented on src/test/java/pages/practicesoftwaretestingpages/LoginPagePST.java:
```diff
@@ -0,0 +1,47 @@
+    @FindBy(xpath = "//div[@data-test='login-error']")
+    private WebElement errorMessage;
+
+    public LoginPagePST(WebDriver driver, WebDriverWaitUtils waitUtils) {
```
as you are initializing the waitUtils in BasePage, you don't need to declare it as parameter in the constructor of the sub-classes

alvaroprado199103 replied:
Fixed

holgiosalos commented on src/test/java/pages/practicesoftwaretestingpages/AccountPage.java:
```diff
@@ -0,0 +1,27 @@
+    @FindBy(xpath = "//a[normalize-space()='Home']")
+    private WebElement homeLink;
+
+    @FindBy(xpath = "//h1[normalize-space()='My account']")
```
I would prefer to use the data-test property and a CSS selector: `h1[data-test='page-title']` as there are two items with the text "My account", you resolved that by specifying the h1 tag in XPath but we should try to use locators that identify the element we want to interact

alvaroprado199103 replied:
Fixed

holgiosalos commented on src/test/java/pages/practicesoftwaretestingpages/AccountPage.java:
```diff
@@ -0,0 +1,27 @@
+import pages.BasePage;
+
+public class AccountPage extends BasePage {
+    @FindBy(xpath = "//a[normalize-space()='Home']")
```
**Optional:** I would use the data-test property and a CSS selector: `a[data-test='nav-home']`
This change is optional as the only element with "Home" text is the link to the home page.

alvaroprado199103 replied:
Fixed

holgiosalos commented on src/test/java/pages/practicesoftwaretestingpages/AccountPage.java:
```diff
@@ -0,0 +1,27 @@
+        PageFactory.initElements(driver, this);
+    }
+
+    public void clickHome() {
```
**Optional:** we should avoid using words like click; we can use goToHome. However in this context it is acceptable to use click

alvaroprado199103 replied:
Fixed

holgiosalos commented on src/test/resources/PracticeSoftwareTesting.feature:
```diff
@@ -0,0 +1,48 @@
+        And the user is authenticated with valid credentials
+
+    Scenario: Successful purchase using combined filters
+        When the user searches for another tool using combined filters
```
We can reword it: `user searches for a tool in the 'Other' category, with a price between '50' and '150', sorting the results by price from 'Low to High'`

alvaroprado199103 replied:
Fixed.

holgiosalos commented on src/test/java/pages/practicesoftwaretestingpages/HomePage.java:
This class has three responsibilities:
1. The related to the navigation bar (go to: login, cart, invoices pages and return to home page) 
2. The logic for filtering products
3. and logic to open the page using the URL 

Please create page objects for responsibilities 1 and 2, then you can include them as part of this page object.
- For responsibility 1, create a page object called NavigationBarComponent
- For responsibility 2, I would name it FilterComponent or SearchOptionsComponent.

alvaroprado199103 replied:
Fixed, created 2 new component clases.

holgiosalos commented on src/test/java/steps/PracticeSoftwareTestingSteps.java:
```diff
@@ -0,0 +1,297 @@
+        Allure.addAttachment("Category Selection", "text/plain", "User selected 'Power Tools' category");
+    }
+
+    @When("the user searches for another tool using combined filters")
```
as mentioned in feature file, we can reword it: `user searches for a tool in the 'Other' category, with a price between '50' and '150', sorting the results by price from 'Low to High'` and receive as params the category, price and sort options

alvaroprado199103 replied:
Fixed.

holgiosalos commented on src/test/resources/PracticeSoftwareTesting.feature:
```diff
@@ -0,0 +1,48 @@
+
+    Scenario: Successful purchase using combined filters
+        When the user searches for another tool using combined filters
+        And adds the third tool to the cart
```
I would use the name of the tool instead of the number

alvaroprado199103 replied:
FIxed.

holgiosalos commented on src/test/java/steps/PracticeSoftwareTestingSteps.java:
```diff
@@ -0,0 +1,297 @@
+        Allure.addAttachment("Add to Cart", "text/plain", "User added 'Cordless Drill 24V' to cart");
+    }
+
+    @And("adds the third tool to the cart")
```
I would use the name of the tool instead of the number

alvaroprado199103 replied:
Fixed.

holgiosalos commented on src/test/java/pages/practicesoftwaretestingpages/FilterPage.java:
```diff
@@ -0,0 +1,100 @@
+    @FindBy(xpath = "//button[@data-test='add-to-cart']")
+    private WebElement addToCartButton;
+
+    public FilterPage(WebDriver driver, WebDriverWaitUtils waitUtils) {
```
as you are initializing the waitUtils in BasePage, you don't need to declare it as parameter in the constructor of the sub-classes

alvaroprado199103 replied:
Fixed

holgiosalos commented on src/test/java/pages/practicesoftwaretestingpages/FilterPage.java:
```diff
@@ -0,0 +1,100 @@
+public class FilterPage extends BasePage {
+    private ApiWaitUtils apiWaitUtils;
+
+    @FindBy(xpath = "(//a[@class='card'])")
```
use css selector instead of xpath for this element

alvaroprado199103 replied:
Fixed.

holgiosalos commented on src/test/java/pages/practicesoftwaretestingpages/FilterPage.java:
```diff
@@ -0,0 +1,100 @@
+    @FindBy(xpath = "(//a[@class='card'])")
+    private List<WebElement> productCards;
+
+    @FindBy(xpath = "(//a[@class='card'])//span[@data-test='product-price']")
```
use css selector instead of xpath for this element, also just "span[data-test='product-price']" should work
![image](https://github.com/user-attachments/assets/3d385e2d-3b3e-4cc1-8bef-864e1d6dac56)

alvaroprado199103 replied:
Fixed.

holgiosalos commented on src/test/java/pages/practicesoftwaretestingpages/FilterPage.java:
```diff
@@ -0,0 +1,100 @@
+    private List<WebElement> productPrices;
+
+    @FindBy(xpath = "(//a[@class='card'])//button[@data-test='add-to-cart']")
+    private List<WebElement> addToCartButtons;
```
This property is not used in the class. remove it

alvaroprado199103 replied:
Fixed

holgiosalos commented on src/test/java/pages/practicesoftwaretestingpages/FilterPage.java:
```diff
@@ -0,0 +1,100 @@
+    @FindBy(xpath = "(//a[@class='card'])//button[@data-test='add-to-cart']")
+    private List<WebElement> addToCartButtons;
+
+    @FindBy(xpath = "(//h5[@class='card-title'])")
```
use css selector instead of xpath, in this case you have two options, use the data-test or use the class.
![image](https://github.com/user-attachments/assets/b08d96cb-be11-4d3e-9f67-efcf4a842e8b)

alvaroprado199103 replied:
Fixed

holgiosalos commented on src/test/java/pages/practicesoftwaretestingpages/FilterPage.java:
```diff
@@ -0,0 +1,100 @@
+    @FindBy(xpath = "(//h5[@class='card-title'])")
+    private List<WebElement> productNames;
+
+    @FindBy(xpath = "//select[@data-test='sort-select']")
+    private WebElement sortDropdown;
+
+    @FindBy(xpath = "//input[@data-test='price-min']")
+    private WebElement minPriceInput;
+
+    @FindBy(xpath = "//input[@data-test='price-max']")
+    private WebElement maxPriceInput;
+
+    @FindBy(xpath = "//button[@data-test='filter-button']")
+    private WebElement filterButton;
```
If I understand this class's purpose well, these elements should not be part of this class, as they are not related to the product list.

alvaroprado199103 replied:
fixed.

holgiosalos commented on src/test/java/pages/practicesoftwaretestingpages/FilterPage.java:
```diff
@@ -0,0 +1,100 @@
+
+import java.util.List;
+
+public class FilterPage extends BasePage {
```
If I understand this class's purpose well, I would name it differently. I would use "ProductListPage" or "ProductsComponent"

alvaroprado199103 replied:
Fixed to ProductListPage

holgiosalos commented on src/test/java/pages/practicesoftwaretestingpages/FilterPage.java:
```diff
@@ -0,0 +1,100 @@
+                .anyMatch(element -> element.getText().equals(productName));
+    }
+
+    public void sortByPrice(String sortOption) {
+        waitUtils.waitForElementToBeClickable(sortDropdown);
+        sortDropdown.click();
+        sortDropdown.sendKeys(sortOption);
+    }
+
+    public void setPriceRange(String minPrice, String maxPrice) {
+        waitUtils.waitForVisibility(minPriceInput);
+        minPriceInput.clear();
+        minPriceInput.sendKeys(minPrice);
+
+        waitUtils.waitForVisibility(maxPriceInput);
+        maxPriceInput.clear();
+        maxPriceInput.sendKeys(maxPrice);
+
+        waitUtils.waitForElementToBeClickable(filterButton);
+        filterButton.click();
+    }
```
If I understand this class's purpose well, these methods should not be part of this class, as they are not related to the product list.

alvaroprado199103 replied:
methods added to the Filtercomponent. java class.

holgiosalos commented on src/test/java/pages/practicesoftwaretestingpages/FilterPage.java:
```diff
@@ -0,0 +1,100 @@
+
+    public double getProductPrice(String productName) {
+        waitUtils.waitForVisibility(productCards.get(0));
+        for (int i = 0; i < productNames.size(); i++) {
+            if (productNames.get(i).getText().equals(productName)) {
+                String priceText = productPrices.get(i).getText().replace("$", "").trim();
+                return Double.parseDouble(priceText);
+            }
+        }
+        return 0.0;
```
You can probably use a stream to perform this logic. Use a logger instead of System.out, and use English instead of Spanish

alvaroprado199103 replied:
method deleted

holgiosalos commented on src/test/java/pages/practicesoftwaretestingpages/FilterPage.java:
```diff
@@ -0,0 +1,100 @@
+
+    public void selectProduct(String productName) {
+        waitUtils.waitForVisibility(productCards.get(0));
+        for (int i = 0; i < productCards.size(); i++) {
+            if (productNames.get(i).getText().equals(productName)) {
+                System.out.println("Select product: Producto encontrado: " + productNames.get(i).getText());
+                productNames.get(i).click();
+                break;
+            }
+        }
```
You can probably use a stream to perform this logic. Use a logger instead of System.out, and use English instead of Spanish

alvaroprado199103 replied:
Fixed

holgiosalos commented on src/test/java/pages/practicesoftwaretestingpages/ProductPage.java:
```diff
@@ -0,0 +1,49 @@
+import pages.BasePage;
+
+public class ProductPage extends BasePage {
+    @FindBy(xpath = "//button[@data-test='add-to-cart']")
+    private WebElement addToCartButton;
+
+    @FindBy(xpath = "//h1[@data-test='product-name']")
+    private WebElement productName;
+
+    @FindBy(xpath = "//span[@data-test='product-price']")
+    private WebElement productPrice;
+
+    @FindBy(xpath = "//div[@data-test='product-description']")
+    private WebElement productDescription;
+
+    @FindBy(xpath = "//div[@id='toast-container']")
+    private WebElement toastMessage;
+
+    @FindBy(xpath = "//a[@data-test='nav-home']")
+    private WebElement homeButton;
```
Use css selectors instead of xpath

alvaroprado199103 replied:
Fixed.

holgiosalos commented on src/test/java/pages/practicesoftwaretestingpages/LoginPagePST.java:
```diff
@@ -0,0 +1,47 @@
+import pages.BasePage;
+
+public class LoginPagePST extends BasePage {
+    @FindBy(xpath = "//input[@id='email']")
+    private WebElement emailInput;
+
+    @FindBy(xpath = "//input[@id='password']")
+    private WebElement passwordInput;
+
+    @FindBy(xpath = "//input[@value='Login']")
+    private WebElement loginButton;
+
+    @FindBy(xpath = "//div[@data-test='login-error']")
+    private WebElement errorMessage;
```
Use CSS selectors instead of xpath

alvaroprado199103 replied:
Fixed.

holgiosalos commented on src/test/java/pages/practicesoftwaretestingpages/HomePage.java:
```diff
@@ -0,0 +1,135 @@
+    @FindBy(xpath = "//a[@data-test='nav-my-invoices']")
+    private WebElement myInvoicesButton;
+
+    public HomePage(WebDriver driver, WebDriverWaitUtils waitUtils) {
```
as you are initializing the waitUtils in BasePage, you don't need to declare it as parameter in the constructor of the sub-classes

alvaroprado199103 replied:
Fixed

holgiosalos commented on src/test/java/pages/practicesoftwaretestingpages/LoginPagePST.java:
```diff
@@ -0,0 +1,47 @@
+    @FindBy(xpath = "//div[@data-test='login-error']")
+    private WebElement errorMessage;
+
+    public LoginPagePST(WebDriver driver, WebDriverWaitUtils waitUtils) {
```
as you are initializing the waitUtils in BasePage, you don't need to declare it as parameter in the constructor of the sub-classes

alvaroprado199103 replied:
Fixed.

holgiosalos commented on src/test/java/pages/practicesoftwaretestingpages/ProductPage.java:
```diff
@@ -0,0 +1,49 @@
+    @FindBy(xpath = "//a[@data-test='nav-home']")
+    private WebElement homeButton;
+
+    public ProductPage(WebDriver driver, WebDriverWaitUtils waitUtils) {
```
as you are initializing the waitUtils in BasePage, you don't need to declare it as parameter in the constructor of the sub-classes

alvaroprado199103 replied:
Fixed.

holgiosalos commented on src/test/java/pages/practicesoftwaretestingpages/LoginPagePST.java:
```diff
@@ -0,0 +1,47 @@
+import pages.BasePage;
+
+public class LoginPagePST extends BasePage {
+    @FindBy(xpath = "//input[@id='email']")
+    private WebElement emailInput;
+
+    @FindBy(xpath = "//input[@id='password']")
+    private WebElement passwordInput;
+
+    @FindBy(xpath = "//input[@value='Login']")
+    private WebElement loginButton;
+
+    @FindBy(xpath = "//div[@data-test='login-error']")
+    private WebElement errorMessage;
```
use css selectors instead of xpath

alvaroprado199103 replied:
Fixed.

holgiosalos commented on src/test/java/pages/practicesoftwaretestingpages/ProductPage.java:
```diff
@@ -0,0 +1,49 @@
+    private WebElement productDescription;
+
+    @FindBy(xpath = "//div[@id='toast-container']")
+    private WebElement toastMessage;
```
use css selectors instead of xpath

alvaroprado199103 replied:
Fixed.

holgiosalos commented on src/test/java/pages/practicesoftwaretestingpages/ProductPage.java:
```diff
@@ -0,0 +1,49 @@
+    @FindBy(xpath = "//div[@id='toast-container']")
+    private WebElement toastMessage;
+
+    @FindBy(xpath = "//a[@data-test='nav-home']")
+    private WebElement homeButton;
```
This should be part of the suggested new page object called NavigationBarComponent, and I would use that object within this class

alvaroprado199103 replied:
Fixed.

holgiosalos commented on src/test/java/pages/practicesoftwaretestingpages/ProductPage.java:
```diff
@@ -0,0 +1,49 @@
+    @FindBy(xpath = "//a[@data-test='nav-home']")
+    private WebElement homeButton;
+
+    public ProductPage(WebDriver driver, WebDriverWaitUtils waitUtils) {
```
as you are initializing the waitUtils in BasePage, you don't need to declare it as parameter in the constructor of the sub-classes

alvaroprado199103 replied:
Fixed.

holgiosalos reviewed (changes requested):
I have not finished my review, but you can work on these comments in the meanwhile

alvaroprado199103 committed:
refactor: centralize navigation in NavigationBarComponent

- Removed goToHome method and homeLink locator from AccountPage
- Removed goToHome method and homeButton locator from ProductPage
- Centralized all navigation in NavigationBarComponent
- Improved code organization following single responsibility principle
- Removed duplicate code
- Changes based on line manager's code review

alvaroprado199103 committed:
refactor: centralize navigation in NavigationBarComponent

alvaroprado199103 committed:
refactor: centralize navigation in NavigationBarComponent

alvaroprado199103 committed:
refactor: centralize navigation in NavigationBarComponent

- Removed goToHome method and homeLink locator from AccountPage
- Removed goToHome method and homeButton locator from ProductPage
- Centralized all navigation in NavigationBarComponent
- Improved code organization following single responsibility principle
- Removed duplicate code
- Changes based on line manager's code review

alvaroprado199103 committed:
refactor: centralize navigation in NavigationBarComponent

alvaroprado199103 committed:
refactor: centralize navigation in NavigationBarComponent

alvaroprado199103 committed:
revert: restore original allure-report.yml workflow

- Removed force parameter from gh-pages action
- Removed Create gh-pages branch step
- Restored original workflow configuration
- Reverted changes that caused deployment issues
